Углубленная разработка туду листа

Tanya-g99 opened the pull request:
closes #53

Tanya-g99 committed:
Внесены изменения для начала работы над редактированием задач

Tanya-g99 commented:
Внесенные изменения:
- Произведено заполнение форм просмотра информации и редактирования данными задачи
 - Добавлена возможность раскрытия и сворачивания списков лейблов в форме создания, в форме редактирования и в сортировке
 - Теперь возможно добавление лейбла в контейнер при редактировании
 - Реализован фильтр задач по категории
 - Исправлена ширина выпадающего списка в форме создания задачи

Tanya-g99 committed:
Добавлена возможность редактировать задачи

geogreck commented on frontend/js/list/task.ts:
```diff
@@ -156,7 +153,16 @@ class Task {
+        this.desc = d
+        this.updateHTML()
+        sendRequest('PATCH', server + `/tasks/${this.id}`, JSON.stringify(this))
+    }
```
Ужасно написанный `PATCH` запрос

geogreck commented on frontend/js/list/todolist.ts:
```diff
@@ -119,18 +119,32 @@ document.addEventListener('click', (e) => {
+        if (buf?.getDueDate()!) {
+            let month = buf?.getDueDate()!.getMonth()!
+            let day = buf?.getDueDate()!.getDate()!
+            dateBoxInfo.value = `${buf?.getDueDate()!.getFullYear()}-${month > 9 ? month : `0${month}`}-${day > 9 ? day : `0${day}`}`
```
Красивый индийский код

geogreck commented on frontend/js/list/todolist.ts:
```diff
@@ -119,18 +119,32 @@ document.addEventListener('click', (e) => {
+            let hours = buf?.getDueDate().getHours()
+            let minutes = buf?.getDueDate().getMinutes()
+            if (hours || minutes) {
+                timeBoxInfo.value = `${hours! > 9 ? hours : `0${hours}`}:${minutes! > 9 ? minutes : `0${minutes}`}`
```
```suggestion
                timeBoxInfo.value = buf?.getDueDate().toTimeString()!.substring(0,5)!
```

geogreck commented on frontend/js/list/todolist.ts:
```diff
@@ -119,18 +119,32 @@ document.addEventListener('click', (e) => {
         let taskLabels = <Array<Label>>buf?.getLabels()
         let container = document.getElementsByClassName('chosen__categories')[1]
-        container.innerHTML = ''
+        taskLabels.length === 0 ? container.innerHTML = 'Категории отсутствуют' : container.innerHTML = ''
```
Открытие задачи без лейблов приведет к ошибке:
```js
Uncaught TypeError: Cannot read properties of undefined (reading 'length')
    at HTMLDocument.<anonymous>
```

Tanya-g99 replied:
Всё знаю, это уже исправлено - ждите коммита

Tanya-g99 committed:
Внесены изменения

Tanya-g99 commented:
Изменения:
- Исправлено редактирование задачи
 - Изменен внешний вид выпадающего списка категорий сортировки для дальнейшего изменения фильтра задач
 - Добавлена прозрачность фона лейблов, чтобы при выборе  темного цвета название лейбла было читаемым

Tanya-g99 committed:
Внесены изменения в список дел

 - Исправлено сворачивание выпадающих списков категорий в формах создания и
 редактирования задачи
 - Улучшено добавление категорий
 - Реализована отправка лейблов на сервер
 - Реализовано архивирование
 - Ограничен доступ к редактированию завершенных и архивированных задач
 - У задач без описания теперь нет иконки его присутствия
 - Архивированные задачи визуально выделены
 - Реализован фильтр одновременно по статусу и выбранным категориям
 - Изменен внешний вид лейблов

Tanya-g99 committed:
Реализован поиск лейблов в списке и произведен рефакторинг: повторяющиеся фрагменты кода вынесены в функции

Tanya-g99 committed:
Внесены изменения для общего меню

geogreck commented on frontend/js/list/todolist.ts:
```diff
@@ -110,42 +128,61 @@ document.addEventListener('click', (e) => {
         titleModal.innerHTML = 'Информация о задаче'
         let btnEditSave = <HTMLInputElement>document.getElementsByClassName('btn-edit-save')[0] //меняем кнопку внизу
         btnEditSave.style.display = 'none'
```
На строке 124 создаётся переменная содержащая модальную форму. Но эта модальная форма уже содержится в глобальной переменной `modalEdit`

Tanya-g99 committed:
Исправлено расстояние между лейблами

Tanya-g99 committed:
Добавлена ссылка на инвентарь
